src/vm/peimagelayout.cpp

```diff
@@ -85,6 +102,21 @@ PEImageLayout* PEImageLayout::Map(HANDLE hFile, PEImage* pOwner)
     RETURN pAlloc.Extract();    
 }
 
+#ifdef FEATURE_PAL
+PEImageLayout* PEImageLayout::MapPreload(LPCWSTR path)
+{
+    void *addr = PAL_malloc(sizeof(MinimalMappedImageLayout));
```

Comment:
Why do you use malloc and placement new instead of just regular `new` everywhere?

Reply:
`new` can't be used before CoreCLR is initialized, and MapPreload is supposed to be called before CoreCLR is initialized. I have updated pull request description with some details

Reply:
You don't need to have coreclr initialized to use PAL functions, you just need to initialize PAL in the coreclr_preload_assembly. Would that cause trouble with the forking? 
Anyways, while calling PAL_malloc works now, calling PAL functions without initializing the PAL is not a good thing. We might decide to add some logging or something that depends on the PAL initialization to that function later.

Reply:
We considered initialization of PAL before coreclr_initialize, but it creates `CPalSynchMgrController`, which prevents forking after preload. Which functions does this thread currently perform? Could it be created later or replaced with something else?

Reply:
This thread manages inter-process synchronization. It seems that we could create and start the thread separately from the other initialization. A reasonable way to do that could be to:
* Add a new PAL initialization method called PAL_InitializeStartThreads. This method would be called after the fork.
* Move the code from inside of `if (flags & PAL_INITIALIZE_SYNC_THREAD)` in the Initialization of PAL to the new method
* In the new method, ensure that the thread starting code could executed only once in a process even if the function is called multiple times, similar to what we do in the Initialize function.
* In the new method, ensure that proper cleanup happens if starting the thread fails.
* Remove the PAL_INITIALIZE_SYNC_THREAD constant definition and usage from pal.h

Reply:
Note that there are number of other places that are not fork-compatible - both in the PAL initialization and CoreCLR initialization. One example from many: `gPID = getpid();`. So this is not just about starting threads.

Reply:
Ah, right.

Reply:
What is a minimal amount of initialization that should happen for PAL in order to use some functions from it like PAL_malloc?

If there is a strict border between two stages of PAL initialization, assert could be added to all PAL function with the condition that appropriate stage of PAL initialization has completed.

For example, there is one more place that creates thread - `SEHInitializeMachExceptions` call in `Initialize` from PAL. Could everything before it be moved to the first stage, and everything after - to the second? Then, PAL_malloc will check that first stage has completed.

There is also a concern that PEImage is used without initialization of CoreCLR. Could the parts related to preloading be moved from CoreCLR and PAL to the separate component, which will be initialized before preloading, and will initialize minimum from PAL?

Reply:
I think it would be best to move the list of pre-loaded images to the PAL. Once it is in the PAL you have much better control over what it calls, and it would require initializing the PAL at all in the ideal case.

Reply:
This is a good idea, we tried to think in this direction, but the problem is that non-PAL version seems to use `MappedImageLayout` too, and we couldn't simply move `MappedImageLayout` to PAL. Is it required for non-PAL version?

Reply:
To clarify, should LoadLibrary load native image and apply relocations, or MappedImageLayout can be used for this?

Reply:
I would avoid touching MappedImageLayout.

Instead:
- Refactor MAPMapPEFile to part that just does the memory mapping and the rest
- Create a clone of `PEImageLayout::ApplyBaseRelocations` in the PAL that works directly against the metal without, and does not need any of the PAL initialized
- Change PAL_LOADLoadPEFile (or introduce a new API) to consult the list of preloaded files first and return it if there is one.